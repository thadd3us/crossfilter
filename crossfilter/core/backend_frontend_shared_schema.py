"""Shared schema definitions for backend-frontend communication."""

from dataclasses import dataclass
from enum import StrEnum
from typing import Any, Optional

from pydantic import BaseModel, Field


class ProjectionType(StrEnum):
    """Types of data projections supported by crossfilter."""
    
    TEMPORAL = "temporal"
    GEO = "geo"
    CLIP_EMBEDDING = "clip_embedding"  # placeholder for future


class FilterOperatorType(StrEnum):
    """Types of filter operations supported by crossfilter."""
    
    INTERSECTION = "intersection"  # Limit to current selection
    SUBTRACTION = "subtraction"    # Remove current selection


@dataclass(frozen=True)
class FilterEvent:
    """Represents a filter event from a specific projection."""
    
    projection_type: ProjectionType
    selected_df_ids: set[int]
    filter_operator: FilterOperatorType


# API Request Models
class LoadDataRequest(BaseModel):
    """Request model for loading data."""
    
    file_path: str


class DfIdsFilterRequest(BaseModel):
    """Request model for filtering to specific df_ids from a plot."""
    
    df_ids: list[int] = Field(..., description="df_ids from the plot (could be from lasso selection, visible area, etc.)")
    event_source: ProjectionType = Field(..., description="Which plot type this filtering comes from")
    filter_operator: FilterOperatorType = Field(..., description="Type of filter operation to apply")


# API Response Models
class SessionStateResponse(BaseModel):
    """Response model for session state information."""
    
    all_rows_count: int
    filtered_rows_count: int
    columns: list[str]
    memory_usage_mb: str
    temporal_projection: dict[str, Any]
    geo_projection: dict[str, Any]
    # Frontend-compatible aliases
    has_data: bool
    row_count: int
    filtered_count: int


class LoadDataResponse(BaseModel):
    """Response model for data loading operations."""
    
    success: bool
    message: str
    session_state: SessionStateResponse


class FilterResponse(BaseModel):
    """Response model for filter operations."""
    
    success: bool
    filter_state: SessionStateResponse


class TemporalPlotResponse(BaseModel):
    """Response model for temporal plot data."""
    
    plotly_plot: dict[str, Any]
    data_type: str
    point_count: int
    distinct_point_count: int
    aggregation_level: Optional[str]


class GeoPlotResponse(BaseModel):
    """Response model for geo plot data."""
    
    plotly_plot: dict[str, Any]
    marker_count: int
    distinct_point_count: int
    aggregation_level: Optional[str]


# Server-Sent Events Models
class SSEEvent(BaseModel):
    """Base model for Server-Sent Events."""
    
    type: str
    timestamp: float
    version: int


class FilterChangeEvent(SSEEvent):
    """Server-Sent Event for filter changes."""
    
    affected_components: list[str]
    session_state: SessionStateResponse


class HeartbeatEvent(SSEEvent):
    """Server-Sent Event for heartbeat."""
    
    pass


class ConnectionEstablishedEvent(FilterChangeEvent):
    """Server-Sent Event for connection establishment."""
    
    pass


