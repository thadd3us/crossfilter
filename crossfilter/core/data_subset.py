from dataclasses import dataclass

import pandas as pd


@dataclass
class DataSubset:
    subset_uuid: str

    df: pd.DataFrame
