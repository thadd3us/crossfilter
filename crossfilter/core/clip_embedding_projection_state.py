"""CLIP embedding projection state management for crossfilter visualization."""

import logging
from typing import Optional

import pandas as pd

from crossfilter.core.bucketing import (
    bucket_by_target_column,
    get_optimal_semantic_embedding_umap_h3_level,
)
from crossfilter.core.schema import get_semantic_embedding_umap_h3_column_name
from crossfilter.core.projection_state import ProjectionState
from crossfilter.core.schema import SchemaColumns as C

logger = logging.getLogger(__name__)


class ClipEmbeddingProjectionState:
    """
    Manages the CLIP embedding projection state for visualizing semantic similarity in 2D space.

    This class handles the spatial aggregation of CLIP embedding data points using H3 hexagonal
    indexing on the UMAP-projected coordinates, maintaining both the current projection DataFrame
    and the aggregation granularity level. The projection may be aggregated at some H3 level (with
    a COUNT column) or may show individual points (no COUNT column).
    """

    def __init__(self, max_rows: int) -> None:
        """
        Initialize CLIP embedding projection state.

        Args:
            max_rows: Maximum number of rows to display before aggregation
        """
        self.projection_state = ProjectionState(max_rows)
        self.current_h3_level: Optional[int] = None

        # User selectable.
        self.max_marker_size: int = 20

    def update_projection(self, filtered_rows: pd.DataFrame) -> None:
        """
        Update the CLIP embedding projection based on the current filtered data.

        Args:
            filtered_rows: Current filtered subset of all_rows
        """
        # Check if semantic embedding UMAP columns are present in the data
        if (
            C.SEMANTIC_EMBEDDING_UMAP_LATITUDE not in filtered_rows.columns
            or C.SEMANTIC_EMBEDDING_UMAP_LONGITUDE not in filtered_rows.columns
        ):
            logger.info(
                "Semantic embedding UMAP columns not found in data, skipping semantic embedding projection update"
            )
            # Set empty projection with no buckets
            self.projection_state.current_bucketing_column = None
            self.current_h3_level = None
            self.projection_state.projection_df = pd.DataFrame()
            return

        # Drop rows with missing semantic embedding UMAP coordinates
        logger.info(
            f"Projecting {len(filtered_rows)=} rows with semantic embedding UMAP coordinates."
        )
        filtered_rows = filtered_rows.dropna(
            subset=[C.SEMANTIC_EMBEDDING_UMAP_LATITUDE, C.SEMANTIC_EMBEDDING_UMAP_LONGITUDE]
        )
        logger.info(f"Kept {len(filtered_rows)=} rows with semantic embedding UMAP coordinates")

        optimal_level = get_optimal_semantic_embedding_umap_h3_level(
            filtered_rows, self.projection_state.max_rows
        )

        if optimal_level is None:
            self.projection_state.current_bucketing_column = None
            self.current_h3_level = None
            self.projection_state.projection_df = filtered_rows
            return

        self.current_h3_level = optimal_level
        self.projection_state.current_bucketing_column = get_semantic_embedding_umap_h3_column_name(
            optimal_level
        )
        self.projection_state.projection_df = bucket_by_target_column(
            filtered_rows,
            self.projection_state.current_bucketing_column,
            self.projection_state.groupby_column,
        )
        logger.info(
            f"Bucketed CLIP embedding data at optimal H3 level {optimal_level=}, {len(self.projection_state.projection_df)=}"
        )

    def get_summary(self) -> dict:
        """Get a summary of the current CLIP embedding projection state."""
        summary = self.projection_state.get_summary()
        # Add CLIP embedding-specific information
        summary["h3_level"] = self.current_h3_level
        summary["target_column"] = self.projection_state.current_bucketing_column
        return summary
