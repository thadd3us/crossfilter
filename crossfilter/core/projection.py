from dataclasses import dataclass
import logging
from typing import Any, Final
import uuid
from crossfilter.core.data_subset import DataSubset
import enum
from crossfilter.core.schema import SchemaColumns as C

import pandas as pd

logger = logging.getLogger(__name__)


# Need to have the same type when comparing NA and present values.
GROUPBY_NA_FILL_VALUE = "?"


class _C(enum.StrEnum):
    GROUP_NAME = "_GROUP_NAME"


def groupby_fillna(df: pd.DataFrame) -> pd.DataFrame:
    """Fill NaN values in the groupby column with a placeholder value."""
    return df.astype(object).fillna(GROUPBY_NA_FILL_VALUE).astype(str)


@dataclass
class _Group:
    group_name: str
    group_df: pd.DataFrame
    unbucketed_row_count: int = -1
    optional_bucketed_on_column: str | None = None

    def __post_init__(self) -> None:
        self.unbucketed_row_count = len(self.group_df)


@dataclass
class GroupedBucketedProjection:
    projection_uuid: str

    # Each group may be bucketed by a different column, depending on how many rows are in the group.
    grouped_by_columns: list[str, ...]

    # Key: how is this indexed?  Can it be (group_key)->bucket_key|uuid?
    projected_data: pd.DataFrame

    name_to_group: dict[str, _Group]


def create_grouped_bucketed_projection(
    ds: DataSubset,
    group_by_columns: list[str, ...],
    # Determines whether we need to bucket.
    max_entries_per_group: int,
    bucketing_candidate_columns: tuple[str, ...],
    sort_groups_by_cardinality: bool,
) -> GroupedBucketedProjection:

    groups: list[_Group] = []
    if group_by_columns:
        df = ds.df.copy()

        df.loc[:, group_by_columns] = groupby_fillna(df.loc[:, group_by_columns])
        df[_C.GROUP_NAME] = df[group_by_columns].apply(
            lambda x: ", ".join(x), axis="columns"
        )

        for group_name, group_df in df.groupby(_C.GROUP_NAME, dropna=False):
            group = _Group(group_name=group_name, group_df=group_df)
            groups.append(group)
    else:
        group = _Group(group_name="", group_df=ds.df)
        groups.append(group)

    for group in groups:
        _maybe_bucket_group(group, max_entries_per_group, bucketing_candidate_columns)

    if sort_groups_by_cardinality:
        groups.sort(key=lambda x: (-x.unbucketed_row_count, x.group_name))
    else:
        groups.sort(key=lambda x: x.group_name)

    result = GroupedBucketedProjection(
        projection_uuid=str(uuid.uuid4()),
        grouped_by_columns=group_by_columns,
        projected_data=pd.concat(
            [group.group_df for group in groups], ignore_index=False
        ),
        name_to_group={g.group_name: g for g in groups},
    )
    return result


def _maybe_bucket_group(
    group: _Group,
    max_entries: int,
    bucketing_candidate_columns: tuple[str, ...],
) -> None:
    df = group.group_df
    original_count = len(df)
    if len(df) <= max_entries:
        logger.info(f"{group.group_name=}: {len(df)=} <= {max_entries=}, not bucketing")
        return
    if not bucketing_candidate_columns:
        logger.warning(
            f"{group.group_name=}: Want to bucket {len(df)=}, but have no columns."
        )
        return

    last_count = len(group.group_df)
    logger.info(f"{group.group_name=}: Looking for bucketing scheme for {len(df)=}")
    for c in bucketing_candidate_columns:
        assert df[c].notna().all()
        this_count = df[c].nunique()
        # This assert could fire because H3 levels are not perfectly hierarchical.
        assert this_count <= last_count, f"{this_count=} > {last_count=} for {c=}"
        if this_count <= max_entries:
            break
    logger.info(
        f"{group.group_name=}: Using bucketing scheme on {c=} with {this_count=}"
    )
    group.optional_bucketed_on_column = c

    df[C.COUNT] = df.groupby(group.optional_bucketed_on_column).transform("size")
    # Keep the first.
    df = df.drop_duplicates(subset=[group.optional_bucketed_on_column])
    assert len(df) == this_count, f"{len(df)=} != {this_count=}, {c=}"
    df = df.reset_index()
    df.index.name = C.DF_ID
    sum_count = df[C.COUNT].sum()
    assert sum_count == original_count, f"{sum_count=} != {original_count=}, {c=}"

    group.group_df = df
