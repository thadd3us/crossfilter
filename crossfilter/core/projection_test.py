from io import StringIO
import pytest
import pandas as pd
from syrupy.assertion import SnapshotAssertion

from crossfilter.core.data_subset import DataSubset
from crossfilter.core.projection import (
    create_grouped_bucketed_projection,
    groupby_fillna,
    _C,
)
from crossfilter.core.schema import DataSchema as C


@pytest.fixture
def ds() -> DataSubset:
    csv_data = f"""{C.UUID_STRING},{C.DATA_TYPE},{C.RATING_0_TO_5},{C.NAME},{C.TIMESTAMP_UTC},TIMESTAMP_YEAR,TIMESTAMP_DAY,H3_01,H3_02
uuid0,GPS,1,name0,2024-07-27T9:00:00,2024-01-01,2024-07-27,abcd,abcd1
uuid1,GPS,2,name1,2024-07-28T9:00:00,2024-01-01,2024-07-28,abcd,abcd1
uuid2,GPS,3,name2,2024-07-29T9:00:00,2024-01-01,2024-07-29,abcd,abcd2
uuid3,GPS,4,name3,2024-07-30T9:00:00,2024-01-01,2024-07-30,abcd,abcd2
uuid4,PHOTO,5,name4,2024-07-31T9:00:00,2024-01-01,2024-07-31,abce,abce1
uuid5,PHOTO,,name5,2025-08-01T9:00:00,2025-01-01,2025-08-01,abce,abce1
uuid6,,,name6,2025-08-02T9:00:00,2025-01-01,2025-08-02,abce,abce2
uuid7,PHOTO,2,name7,2025-08-03T9:00:00,2025-01-01,2025-08-03,abce,abce2
uuid8,VIDEO,3,name8,2025-08-04T9:00:00,2025-01-01,2025-08-04,abce,abce3
uuid9,,4,name9,,,,abcd,abcde 
uuid10,VIDEO,5,name10,2025-08-06T9:00:00,2025-01-01,2025-08-06,, 
"""
    df = pd.read_csv(StringIO(csv_data))
    df[C.RATING_0_TO_5] = df[C.RATING_0_TO_5].astype(pd.Int32Dtype())
    # df = df.set_index(C.UUID_STRING, verify_integrity=True)
    return DataSubset(subset_uuid="test_subset_uuid", df=df)


def test_create_grouped_bucketed_projection_no_grouping_no_bucketing(
    ds: DataSubset,
    snapshot: SnapshotAssertion,
) -> None:
    projection = create_grouped_bucketed_projection(
        ds,
        group_by_columns=[],
        bucketing_candidate_columns=(),
        max_entries_per_group=100,
        sort_groups_by_cardinality=False,
    )
    assert projection.projected_data.to_dict(index=True, orient="records") == snapshot


def test_create_grouped_bucketed_projection_group_by_data_type_no_bucketing(
    ds: DataSubset,
    snapshot: SnapshotAssertion,
) -> None:
    projection = create_grouped_bucketed_projection(
        ds,
        group_by_columns=[C.DATA_TYPE],
        bucketing_candidate_columns=(),
        max_entries_per_group=100,
        sort_groups_by_cardinality=False,
    )
    assert projection.projected_data.to_dict(index=True, orient="records") == snapshot
    assert projection.projected_data[_C.GROUP_NAME].value_counts(
        dropna=False
    ).to_dict() == {"GPS": 4, "PHOTO": 3, "?": 2, "VIDEO": 2}


def test_create_grouped_bucketed_projection_group_by_data_type_bucketing_H3_COARSE(
    ds: DataSubset,
    snapshot: SnapshotAssertion,
) -> None:
    ds.df = ds.df.dropna(subset=["H3_01", "H3_02"])
    projection = create_grouped_bucketed_projection(
        ds,
        group_by_columns=[C.DATA_TYPE],
        bucketing_candidate_columns=("H3_02", "H3_01"),
        max_entries_per_group=1,
        sort_groups_by_cardinality=False,
    )
    assert projection.projected_data.to_dict(index=True, orient="records") == snapshot
    name_to_bucket_level = {
        g.group_name: g.optional_bucketed_on_column
        for g in projection.name_to_group.values()
    }
    assert name_to_bucket_level == {
        "?": "H3_01",
        "GPS": "H3_01",
        "PHOTO": "H3_01",
        "VIDEO": None,
    }

    projection = create_grouped_bucketed_projection(
        ds,
        group_by_columns=[C.DATA_TYPE],
        bucketing_candidate_columns=("H3_02", "H3_01"),
        max_entries_per_group=2,
        sort_groups_by_cardinality=False,
    )
    assert projection.projected_data.to_dict(index=True, orient="records") == snapshot
    name_to_bucket_level = {
        g.group_name: g.optional_bucketed_on_column
        for g in projection.name_to_group.values()
    }
    assert name_to_bucket_level == {
        "?": None,
        "GPS": "H3_02",
        "PHOTO": "H3_02",
        "VIDEO": None,
    }


def test_groupby_fillna(snapshot: SnapshotAssertion) -> None:
    df = pd.DataFrame({"A": [None, 2, 3], "B": [4, 5, None], "C": ["7", 8, None]})
    df["A"] = df["A"].astype(pd.Int64Dtype())
    new_df = groupby_fillna(df)
    assert new_df.to_dict(orient="records") == snapshot
    assert new_df["A"].tolist() == ["?", "2", "3"], "Don't be floats!"
