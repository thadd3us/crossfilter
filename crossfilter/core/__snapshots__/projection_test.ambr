# serializer version: 1
# name: test_create_grouped_bucketed_projection_group_by_data_type_bucketing_H3_COARSE
  list([
    dict({
      <SchemaColumns.COUNT: 'COUNT'>: 1.0,
      'DATA_TYPE': '?',
      'H3_01': 'abce',
      'H3_02': 'abce2',
      'NAME': 'name6',
      'RATING_0_TO_5': None,
      'TIMESTAMP_DAY': '2025-08-02',
      'TIMESTAMP_UTC': '2025-08-02T9:00:00',
      'TIMESTAMP_YEAR': '2025-01-01',
      'UUID_STRING': 'uuid6',
      <_C.GROUP_NAME: '_GROUP_NAME'>: '?',
    }),
    dict({
      <SchemaColumns.COUNT: 'COUNT'>: 1.0,
      'DATA_TYPE': '?',
      'H3_01': 'abcd',
      'H3_02': 'abcde ',
      'NAME': 'name9',
      'RATING_0_TO_5': 4,
      'TIMESTAMP_DAY': nan,
      'TIMESTAMP_UTC': nan,
      'TIMESTAMP_YEAR': nan,
      'UUID_STRING': 'uuid9',
      <_C.GROUP_NAME: '_GROUP_NAME'>: '?',
    }),
    dict({
      <SchemaColumns.COUNT: 'COUNT'>: 4.0,
      'DATA_TYPE': 'GPS',
      'H3_01': 'abcd',
      'H3_02': 'abcd1',
      'NAME': 'name0',
      'RATING_0_TO_5': 1,
      'TIMESTAMP_DAY': '2024-07-27',
      'TIMESTAMP_UTC': '2024-07-27T9:00:00',
      'TIMESTAMP_YEAR': '2024-01-01',
      'UUID_STRING': 'uuid0',
      <_C.GROUP_NAME: '_GROUP_NAME'>: 'GPS',
    }),
    dict({
      <SchemaColumns.COUNT: 'COUNT'>: 3.0,
      'DATA_TYPE': 'PHOTO',
      'H3_01': 'abce',
      'H3_02': 'abce1',
      'NAME': 'name4',
      'RATING_0_TO_5': 5,
      'TIMESTAMP_DAY': '2024-07-31',
      'TIMESTAMP_UTC': '2024-07-31T9:00:00',
      'TIMESTAMP_YEAR': '2024-01-01',
      'UUID_STRING': 'uuid4',
      <_C.GROUP_NAME: '_GROUP_NAME'>: 'PHOTO',
    }),
    dict({
      <SchemaColumns.COUNT: 'COUNT'>: nan,
      'DATA_TYPE': 'VIDEO',
      'H3_01': 'abce',
      'H3_02': 'abce3',
      'NAME': 'name8',
      'RATING_0_TO_5': 3,
      'TIMESTAMP_DAY': '2025-08-04',
      'TIMESTAMP_UTC': '2025-08-04T9:00:00',
      'TIMESTAMP_YEAR': '2025-01-01',
      'UUID_STRING': 'uuid8',
      <_C.GROUP_NAME: '_GROUP_NAME'>: 'VIDEO',
    }),
  ])
# ---
# name: test_create_grouped_bucketed_projection_group_by_data_type_bucketing_H3_COARSE.1
  list([
    dict({
      <SchemaColumns.COUNT: 'COUNT'>: nan,
      'DATA_TYPE': '?',
      'H3_01': 'abce',
      'H3_02': 'abce2',
      'NAME': 'name6',
      'RATING_0_TO_5': None,
      'TIMESTAMP_DAY': '2025-08-02',
      'TIMESTAMP_UTC': '2025-08-02T9:00:00',
      'TIMESTAMP_YEAR': '2025-01-01',
      'UUID_STRING': 'uuid6',
      <_C.GROUP_NAME: '_GROUP_NAME'>: '?',
    }),
    dict({
      <SchemaColumns.COUNT: 'COUNT'>: nan,
      'DATA_TYPE': '?',
      'H3_01': 'abcd',
      'H3_02': 'abcde ',
      'NAME': 'name9',
      'RATING_0_TO_5': 4,
      'TIMESTAMP_DAY': nan,
      'TIMESTAMP_UTC': nan,
      'TIMESTAMP_YEAR': nan,
      'UUID_STRING': 'uuid9',
      <_C.GROUP_NAME: '_GROUP_NAME'>: '?',
    }),
    dict({
      <SchemaColumns.COUNT: 'COUNT'>: 2.0,
      'DATA_TYPE': 'GPS',
      'H3_01': 'abcd',
      'H3_02': 'abcd1',
      'NAME': 'name0',
      'RATING_0_TO_5': 1,
      'TIMESTAMP_DAY': '2024-07-27',
      'TIMESTAMP_UTC': '2024-07-27T9:00:00',
      'TIMESTAMP_YEAR': '2024-01-01',
      'UUID_STRING': 'uuid0',
      <_C.GROUP_NAME: '_GROUP_NAME'>: 'GPS',
    }),
    dict({
      <SchemaColumns.COUNT: 'COUNT'>: 2.0,
      'DATA_TYPE': 'GPS',
      'H3_01': 'abcd',
      'H3_02': 'abcd2',
      'NAME': 'name2',
      'RATING_0_TO_5': 3,
      'TIMESTAMP_DAY': '2024-07-29',
      'TIMESTAMP_UTC': '2024-07-29T9:00:00',
      'TIMESTAMP_YEAR': '2024-01-01',
      'UUID_STRING': 'uuid2',
      <_C.GROUP_NAME: '_GROUP_NAME'>: 'GPS',
    }),
    dict({
      <SchemaColumns.COUNT: 'COUNT'>: 2.0,
      'DATA_TYPE': 'PHOTO',
      'H3_01': 'abce',
      'H3_02': 'abce1',
      'NAME': 'name4',
      'RATING_0_TO_5': 5,
      'TIMESTAMP_DAY': '2024-07-31',
      'TIMESTAMP_UTC': '2024-07-31T9:00:00',
      'TIMESTAMP_YEAR': '2024-01-01',
      'UUID_STRING': 'uuid4',
      <_C.GROUP_NAME: '_GROUP_NAME'>: 'PHOTO',
    }),
    dict({
      <SchemaColumns.COUNT: 'COUNT'>: 1.0,
      'DATA_TYPE': 'PHOTO',
      'H3_01': 'abce',
      'H3_02': 'abce2',
      'NAME': 'name7',
      'RATING_0_TO_5': 2,
      'TIMESTAMP_DAY': '2025-08-03',
      'TIMESTAMP_UTC': '2025-08-03T9:00:00',
      'TIMESTAMP_YEAR': '2025-01-01',
      'UUID_STRING': 'uuid7',
      <_C.GROUP_NAME: '_GROUP_NAME'>: 'PHOTO',
    }),
    dict({
      <SchemaColumns.COUNT: 'COUNT'>: nan,
      'DATA_TYPE': 'VIDEO',
      'H3_01': 'abce',
      'H3_02': 'abce3',
      'NAME': 'name8',
      'RATING_0_TO_5': 3,
      'TIMESTAMP_DAY': '2025-08-04',
      'TIMESTAMP_UTC': '2025-08-04T9:00:00',
      'TIMESTAMP_YEAR': '2025-01-01',
      'UUID_STRING': 'uuid8',
      <_C.GROUP_NAME: '_GROUP_NAME'>: 'VIDEO',
    }),
  ])
# ---
# name: test_create_grouped_bucketed_projection_group_by_data_type_no_bucketing
  list([
    dict({
      'DATA_TYPE': '?',
      'H3_01': 'abce',
      'H3_02': 'abce2',
      'NAME': 'name6',
      'RATING_0_TO_5': None,
      'TIMESTAMP_DAY': '2025-08-02',
      'TIMESTAMP_UTC': '2025-08-02T9:00:00',
      'TIMESTAMP_YEAR': '2025-01-01',
      'UUID_STRING': 'uuid6',
      <_C.GROUP_NAME: '_GROUP_NAME'>: '?',
    }),
    dict({
      'DATA_TYPE': '?',
      'H3_01': 'abcd',
      'H3_02': 'abcde ',
      'NAME': 'name9',
      'RATING_0_TO_5': 4,
      'TIMESTAMP_DAY': nan,
      'TIMESTAMP_UTC': nan,
      'TIMESTAMP_YEAR': nan,
      'UUID_STRING': 'uuid9',
      <_C.GROUP_NAME: '_GROUP_NAME'>: '?',
    }),
    dict({
      'DATA_TYPE': 'GPS',
      'H3_01': 'abcd',
      'H3_02': 'abcd1',
      'NAME': 'name0',
      'RATING_0_TO_5': 1,
      'TIMESTAMP_DAY': '2024-07-27',
      'TIMESTAMP_UTC': '2024-07-27T9:00:00',
      'TIMESTAMP_YEAR': '2024-01-01',
      'UUID_STRING': 'uuid0',
      <_C.GROUP_NAME: '_GROUP_NAME'>: 'GPS',
    }),
    dict({
      'DATA_TYPE': 'GPS',
      'H3_01': 'abcd',
      'H3_02': 'abcd1',
      'NAME': 'name1',
      'RATING_0_TO_5': 2,
      'TIMESTAMP_DAY': '2024-07-28',
      'TIMESTAMP_UTC': '2024-07-28T9:00:00',
      'TIMESTAMP_YEAR': '2024-01-01',
      'UUID_STRING': 'uuid1',
      <_C.GROUP_NAME: '_GROUP_NAME'>: 'GPS',
    }),
    dict({
      'DATA_TYPE': 'GPS',
      'H3_01': 'abcd',
      'H3_02': 'abcd2',
      'NAME': 'name2',
      'RATING_0_TO_5': 3,
      'TIMESTAMP_DAY': '2024-07-29',
      'TIMESTAMP_UTC': '2024-07-29T9:00:00',
      'TIMESTAMP_YEAR': '2024-01-01',
      'UUID_STRING': 'uuid2',
      <_C.GROUP_NAME: '_GROUP_NAME'>: 'GPS',
    }),
    dict({
      'DATA_TYPE': 'GPS',
      'H3_01': 'abcd',
      'H3_02': 'abcd2',
      'NAME': 'name3',
      'RATING_0_TO_5': 4,
      'TIMESTAMP_DAY': '2024-07-30',
      'TIMESTAMP_UTC': '2024-07-30T9:00:00',
      'TIMESTAMP_YEAR': '2024-01-01',
      'UUID_STRING': 'uuid3',
      <_C.GROUP_NAME: '_GROUP_NAME'>: 'GPS',
    }),
    dict({
      'DATA_TYPE': 'PHOTO',
      'H3_01': 'abce',
      'H3_02': 'abce1',
      'NAME': 'name4',
      'RATING_0_TO_5': 5,
      'TIMESTAMP_DAY': '2024-07-31',
      'TIMESTAMP_UTC': '2024-07-31T9:00:00',
      'TIMESTAMP_YEAR': '2024-01-01',
      'UUID_STRING': 'uuid4',
      <_C.GROUP_NAME: '_GROUP_NAME'>: 'PHOTO',
    }),
    dict({
      'DATA_TYPE': 'PHOTO',
      'H3_01': 'abce',
      'H3_02': 'abce1',
      'NAME': 'name5',
      'RATING_0_TO_5': None,
      'TIMESTAMP_DAY': '2025-08-01',
      'TIMESTAMP_UTC': '2025-08-01T9:00:00',
      'TIMESTAMP_YEAR': '2025-01-01',
      'UUID_STRING': 'uuid5',
      <_C.GROUP_NAME: '_GROUP_NAME'>: 'PHOTO',
    }),
    dict({
      'DATA_TYPE': 'PHOTO',
      'H3_01': 'abce',
      'H3_02': 'abce2',
      'NAME': 'name7',
      'RATING_0_TO_5': 2,
      'TIMESTAMP_DAY': '2025-08-03',
      'TIMESTAMP_UTC': '2025-08-03T9:00:00',
      'TIMESTAMP_YEAR': '2025-01-01',
      'UUID_STRING': 'uuid7',
      <_C.GROUP_NAME: '_GROUP_NAME'>: 'PHOTO',
    }),
    dict({
      'DATA_TYPE': 'VIDEO',
      'H3_01': 'abce',
      'H3_02': 'abce3',
      'NAME': 'name8',
      'RATING_0_TO_5': 3,
      'TIMESTAMP_DAY': '2025-08-04',
      'TIMESTAMP_UTC': '2025-08-04T9:00:00',
      'TIMESTAMP_YEAR': '2025-01-01',
      'UUID_STRING': 'uuid8',
      <_C.GROUP_NAME: '_GROUP_NAME'>: 'VIDEO',
    }),
    dict({
      'DATA_TYPE': 'VIDEO',
      'H3_01': nan,
      'H3_02': ' ',
      'NAME': 'name10',
      'RATING_0_TO_5': 5,
      'TIMESTAMP_DAY': '2025-08-06',
      'TIMESTAMP_UTC': '2025-08-06T9:00:00',
      'TIMESTAMP_YEAR': '2025-01-01',
      'UUID_STRING': 'uuid10',
      <_C.GROUP_NAME: '_GROUP_NAME'>: 'VIDEO',
    }),
  ])
# ---
# name: test_create_grouped_bucketed_projection_no_grouping_no_bucketing
  list([
    dict({
      'DATA_TYPE': 'GPS',
      'H3_01': 'abcd',
      'H3_02': 'abcd1',
      'NAME': 'name0',
      'RATING_0_TO_5': 1,
      'TIMESTAMP_DAY': '2024-07-27',
      'TIMESTAMP_UTC': '2024-07-27T9:00:00',
      'TIMESTAMP_YEAR': '2024-01-01',
      'UUID_STRING': 'uuid0',
    }),
    dict({
      'DATA_TYPE': 'GPS',
      'H3_01': 'abcd',
      'H3_02': 'abcd1',
      'NAME': 'name1',
      'RATING_0_TO_5': 2,
      'TIMESTAMP_DAY': '2024-07-28',
      'TIMESTAMP_UTC': '2024-07-28T9:00:00',
      'TIMESTAMP_YEAR': '2024-01-01',
      'UUID_STRING': 'uuid1',
    }),
    dict({
      'DATA_TYPE': 'GPS',
      'H3_01': 'abcd',
      'H3_02': 'abcd2',
      'NAME': 'name2',
      'RATING_0_TO_5': 3,
      'TIMESTAMP_DAY': '2024-07-29',
      'TIMESTAMP_UTC': '2024-07-29T9:00:00',
      'TIMESTAMP_YEAR': '2024-01-01',
      'UUID_STRING': 'uuid2',
    }),
    dict({
      'DATA_TYPE': 'GPS',
      'H3_01': 'abcd',
      'H3_02': 'abcd2',
      'NAME': 'name3',
      'RATING_0_TO_5': 4,
      'TIMESTAMP_DAY': '2024-07-30',
      'TIMESTAMP_UTC': '2024-07-30T9:00:00',
      'TIMESTAMP_YEAR': '2024-01-01',
      'UUID_STRING': 'uuid3',
    }),
    dict({
      'DATA_TYPE': 'PHOTO',
      'H3_01': 'abce',
      'H3_02': 'abce1',
      'NAME': 'name4',
      'RATING_0_TO_5': 5,
      'TIMESTAMP_DAY': '2024-07-31',
      'TIMESTAMP_UTC': '2024-07-31T9:00:00',
      'TIMESTAMP_YEAR': '2024-01-01',
      'UUID_STRING': 'uuid4',
    }),
    dict({
      'DATA_TYPE': 'PHOTO',
      'H3_01': 'abce',
      'H3_02': 'abce1',
      'NAME': 'name5',
      'RATING_0_TO_5': None,
      'TIMESTAMP_DAY': '2025-08-01',
      'TIMESTAMP_UTC': '2025-08-01T9:00:00',
      'TIMESTAMP_YEAR': '2025-01-01',
      'UUID_STRING': 'uuid5',
    }),
    dict({
      'DATA_TYPE': nan,
      'H3_01': 'abce',
      'H3_02': 'abce2',
      'NAME': 'name6',
      'RATING_0_TO_5': None,
      'TIMESTAMP_DAY': '2025-08-02',
      'TIMESTAMP_UTC': '2025-08-02T9:00:00',
      'TIMESTAMP_YEAR': '2025-01-01',
      'UUID_STRING': 'uuid6',
    }),
    dict({
      'DATA_TYPE': 'PHOTO',
      'H3_01': 'abce',
      'H3_02': 'abce2',
      'NAME': 'name7',
      'RATING_0_TO_5': 2,
      'TIMESTAMP_DAY': '2025-08-03',
      'TIMESTAMP_UTC': '2025-08-03T9:00:00',
      'TIMESTAMP_YEAR': '2025-01-01',
      'UUID_STRING': 'uuid7',
    }),
    dict({
      'DATA_TYPE': 'VIDEO',
      'H3_01': 'abce',
      'H3_02': 'abce3',
      'NAME': 'name8',
      'RATING_0_TO_5': 3,
      'TIMESTAMP_DAY': '2025-08-04',
      'TIMESTAMP_UTC': '2025-08-04T9:00:00',
      'TIMESTAMP_YEAR': '2025-01-01',
      'UUID_STRING': 'uuid8',
    }),
    dict({
      'DATA_TYPE': nan,
      'H3_01': 'abcd',
      'H3_02': 'abcde ',
      'NAME': 'name9',
      'RATING_0_TO_5': 4,
      'TIMESTAMP_DAY': nan,
      'TIMESTAMP_UTC': nan,
      'TIMESTAMP_YEAR': nan,
      'UUID_STRING': 'uuid9',
    }),
    dict({
      'DATA_TYPE': 'VIDEO',
      'H3_01': nan,
      'H3_02': ' ',
      'NAME': 'name10',
      'RATING_0_TO_5': 5,
      'TIMESTAMP_DAY': '2025-08-06',
      'TIMESTAMP_UTC': '2025-08-06T9:00:00',
      'TIMESTAMP_YEAR': '2025-01-01',
      'UUID_STRING': 'uuid10',
    }),
  ])
# ---
# name: test_groupby_fillna
  list([
    dict({
      'A': '?',
      'B': '4.0',
      'C': '7',
    }),
    dict({
      'A': '2',
      'B': '5.0',
      'C': '8',
    }),
    dict({
      'A': '3',
      'B': '?',
      'C': '?',
    }),
  ])
# ---
# name: test_groupby_fillna.1
  True
# ---
